File: di-ipv-cimit-stub/lambdas/stub-management/src/main/java/uk/gov/di/ipv/core/stubmanagement/StubManagementHandler.java
package uk.gov.di.ipv.core.stubmanagement;

import com.amazonaws.HttpMethod;
import com.amazonaws.services.lambda.runtime.Context;
import com.amazonaws.services.lambda.runtime.RequestHandler;
import com.amazonaws.services.lambda.runtime.events.APIGatewayProxyRequestEvent;
import com.amazonaws.services.lambda.runtime.events.APIGatewayProxyResponseEvent;
import com.fasterxml.jackson.databind.ObjectMapper;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import uk.gov.di.ipv.core.library.model.UserCisRequest;
import uk.gov.di.ipv.core.library.model.UserMitigationRequest;
import uk.gov.di.ipv.core.library.service.CimitStubItemService;
import uk.gov.di.ipv.core.library.service.ConfigService;
import uk.gov.di.ipv.core.library.service.PendingMitigationService;
import uk.gov.di.ipv.core.stubmanagement.exceptions.BadRequestException;
import uk.gov.di.ipv.core.stubmanagement.exceptions.DataNotFoundException;
import uk.gov.di.ipv.core.stubmanagement.service.UserService;

import java.io.IOException;
import java.net.URLDecoder;
import java.nio.charset.StandardCharsets;
import java.util.List;
import java.util.Map;
import java.util.regex.Pattern;

public class StubManagementHandler
        implements RequestHandler<APIGatewayProxyRequestEvent, APIGatewayProxyResponseEvent> {
    private static final Logger LOGGER = LogManager.getLogger();
    private static final ObjectMapper objectMapper = new ObjectMapper();
    private final UserService userService;
    private final PendingMitigationService pendingMitigationService;
    private final CimitStubItemService cimitStubItemService;

    private static final Pattern CIS_PATTERN = Pattern.compile("^/user/[-a-zA-Z0-9_:]+/cis$");
    private static final Pattern CIS_MITIGATIONS =
            Pattern.compile("^/user/[-a-zA-Z0-9_:]+/mitigations/[-a-zA-Z0-9_]+$");
    private static final List<String> SUPPORTED_MITIGATION_METHODS =
            List.of(HttpMethod.POST.toString(), HttpMethod.PUT.toString());

    private static final String USER_ID_PATH_PARAMS = "userId";
    private static final String CI_PATH_PARAMS = "ci";

    public StubManagementHandler() {
        this.userService = new UserService();
        ConfigService configService = new ConfigService();
        this.pendingMitigationService = new PendingMitigationService(configService);
        this.cimitStubItemService = new CimitStubItemService(configService);
    }

    public StubManagementHandler(
            UserService userService,
            PendingMitigationService pendingMitigationService,
            CimitStubItemService cimitStubItemService) {
        this.userService = userService;
        this.pendingMitigationService = pendingMitigationService;
        this.cimitStubItemService = cimitStubItemService;
    }

    @Override
    public APIGatewayProxyResponseEvent handleRequest(
            APIGatewayProxyRequestEvent event, Context context) {
        String httpMethod = event.getHttpMethod();
        try {
            String path = URLDecoder.decode(event.getPath(), StandardCharsets.UTF_8);
            LOGGER.info("Received '{}' event with path '{}'", httpMethod, path);
            Map<String, String> pathParameters = event.getPathParameters();
            String userId =
                    URLDecoder.decode(
                            pathParameters.get(USER_ID_PATH_PARAMS), StandardCharsets.UTF_8);
            if (CIS_PATTERN.matcher(path).matches()) {
                List<UserCisRequest> userCisRequests =
                        objectMapper.readValue(
                                event.getBody(),
                                objectMapper
                                        .getTypeFactory()
                                        .constructCollectionType(List.class, UserCisRequest.class));
                if (httpMethod.equals(HttpMethod.POST.toString())) {
                    userService.addUserCis(userId, userCisRequests);
                } else if (httpMethod.equals(HttpMethod.PUT.toString())) {
                    userService.updateUserCis(userId, userCisRequests);
                } else {
                    return buildErrorResponse("Http Method is not supported.", 400);
                }
            } else if (CIS_MITIGATIONS.matcher(path).matches()) {
                String ci = pathParameters.get(CI_PATH_PARAMS);
                UserMitigationRequest userMitigationRequest =
                        objectMapper.readValue(event.getBody(), UserMitigationRequest.class);
                if (SUPPORTED_MITIGATION_METHODS.contains(httpMethod)) {
                    if (cimitStubItemService.getCiForUserId(userId, ci) == null) {
                        throw new DataNotFoundException("User and ContraIndicator not found.");
                    }
                    pendingMitigationService.persistPendingMitigation(
                            userMitigationRequest, ci, httpMethod);
                } else {
                    return buildErrorResponse("Http Method is not supported.", 400);
                }
            } else {
                return buildErrorResponse("Invalid URI.", 400);
            }
            return buildSuccessResponse();
        } catch (IOException e) {
            LOGGER.error(String.format("IOException : %s", e.getMessage()));
            return buildErrorResponse("Invalid request body.", 400);
        } catch (BadRequestException e) {
            LOGGER.error(e.getMessage());
            return buildErrorResponse(e.getMessage(), 400);
        } catch (DataNotFoundException e) {
            return buildErrorResponse(e.getMessage(), 404);
        } catch (Exception e) {
            LOGGER.info(String.format("Exception : %s", e.getMessage()));
            return buildErrorResponse(e.getMessage(), 500);
        }
    }

    private APIGatewayProxyResponseEvent buildSuccessResponse() {
        return new APIGatewayProxyResponseEvent().withStatusCode(200).withBody("success");
    }

    private APIGatewayProxyResponseEvent buildErrorResponse(String message, int statusCode) {
        return new APIGatewayProxyResponseEvent().withStatusCode(statusCode).withBody(message);
    }
}

File: di-ipv-cimit-stub/lambdas/src/external-api/stub-management/stubManagementHandler.ts
import { APIGatewayProxyEvent, APIGatewayProxyResult } from "aws-lambda";
import {
  addUserCis,
  updateUserCis,
  UserCisRequest,
} from "./service/userService";
import * as cimitStubItemService from "../../common/cimitStubItemService";
import * as pendingMitigationService from "../../common/pendingMitigationService";
import * as preMitigationService from "../../common/preMitigationService";
import { UserMitigationRequest } from "../../common/pendingMitigationService";
import { UserPreMitigationRequest } from "../../common/preMitigationService";

const CIS_PATTERN = /^\/user\/[-a-zA-Z0-9_:]+\/cis$/;
const CIS_MITIGATIONS =
  /^\/user\/[-a-zA-Z0-9_:]+\/mitigations\/[-a-zA-Z0-9_]+$/;
const PRE_MITIGATIONS_PATTERN =
  /^\/user\/[-a-zA-Z0-9_:]+\/premitigations\/[-a-zA-Z0-9_]+$/;
const SUPPORTED_HTTP_METHODS = ["POST", "PUT"];

export const handler = async (
  event: APIGatewayProxyEvent,
): Promise<APIGatewayProxyResult> => {
  const httpMethod = event.httpMethod;

  try {
    const path = decodeURIComponent(event.path);
    console.info(`Received '${httpMethod}' event with path '${path}'`);

    if (!SUPPORTED_HTTP_METHODS.includes(httpMethod)) {
      return buildErrorResponse("Http Method is not supported.", 400);
    }

    const pathParameters = event.pathParameters || {};
    const userId = decodeURIComponent(pathParameters.userId || "");

    if (CIS_PATTERN.test(path)) {
      let userCisRequests: UserCisRequest[];
      try {
        userCisRequests = JSON.parse(event.body || "");
      } catch {
        return buildErrorResponse("Invalid request body", 400);
      }
      if (
        userCisRequests.some(
          (request) => !request.code || request.code.trim() === "",
        )
      ) {
        return buildErrorResponse("CI codes cannot be empty.", 400);
      }

      if (httpMethod === "POST") {
        await addUserCis(userId, userCisRequests);
      } else if (httpMethod === "PUT") {
        await updateUserCis(userId, userCisRequests);
      }
    } else if (CIS_MITIGATIONS.test(path)) {
      const ci = pathParameters.ci || "";
      let userMitigationRequest: UserMitigationRequest;
      try {
        userMitigationRequest = JSON.parse(event.body || "");
      } catch {
        return buildErrorResponse("Invalid request body", 400);
      }

      const ciData = await cimitStubItemService.getCiForUserId(userId, ci);
      if (!ciData || ciData.length === 0) {
        return buildErrorResponse("User and ContraIndicator not found.", 404);
      }
      await pendingMitigationService.persistPendingMitigation(
        userMitigationRequest,
        ci,
        httpMethod,
      );
    } else if (PRE_MITIGATIONS_PATTERN.test(path)) {
      const ci = pathParameters.ci || "";

      let userPreMitigationRequest: UserPreMitigationRequest;
      try {
        userPreMitigationRequest = JSON.parse(event.body || "");
      } catch {
        return buildErrorResponse("Invalid request body", 400);
      }

      await preMitigationService.persistPreMitigation(
        userId,
        ci,
        userPreMitigationRequest,
      );
    } else {
      return buildErrorResponse("Invalid URI.", 400);
    }

    return buildSuccessResponse();
  } catch (error: unknown) {
    console.info(`Unexpected error : ${error}`);
    return buildErrorResponse(`Unexpected error : ${error}`, 500);
  }
};

function buildSuccessResponse(): APIGatewayProxyResult {
  return { statusCode: 200, body: "success" };
}

function buildErrorResponse(
  message: string,
  statusCode: number,
): APIGatewayProxyResult {
  return { statusCode, body: message };
}
